crypto/modexp/gmp/gmp.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package gmp
+
+// #cgo LDFLAGS: -lgmp
+// #include "modexp.h"
+import "C"
+import (
+	"errors"
+	"runtime"
+	"unsafe"
+)
+
+// ModExp performs modular exponentiation using GMP
+// This is thread safe.
+func ModExp(base, exp, mod []byte) ([]byte, error) {
+	// Handle empty modulus - return empty result (EVM behavior)
+	if len(mod) == 0 {
+		return []byte{}, nil
+	}
+
+	// Special case: zero modulus
+	allZero := true
+	for _, b := range mod {
+		if b != 0 {
+			allZero = false
+			break
+		}
+	}
+	if allZero {
+		return []byte{}, nil
+	}
```

Comment:
Haven't checked how GMP handles modulus == 0 -- if it does, then can remove this if its faster for gmp to do it